File: leetcode_cn/82.py
# %%
from typing import List


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def generate(vals: List[int]) -> ListNode:
    ans = ListNode(-1, None)
    tail = ans
    for x in vals:
        t = ListNode(x, None)
        tail.next = t
        tail = t
    return ans.next


def show(head: ListNode) -> List[int]:
    ans = list()
    h = head
    while h:
        ans.append(h.val)
        h = h.next
    return ans


class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        ans = ListNode(-1, None)
        h = head
        tail = ans
        while h:
            if h.next and h.val == h.next.val:  # 存在重复元素
                # 找到重复的下一个
                while h.next and h.val == h.next.val:
                    h = h.next
            else:
                t = ListNode(h.val, None)
                tail.next = t
                tail = t
            h = h.next
        return ans.next


headList = []
headNode = generate(headList)
so = Solution()
ans = so.deleteDuplicates(headNode)
ansList = show(ans)
print(ansList)

File: leetcode_cn/82.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
		ListNode *dummy = new ListNode(0,head);
		ListNode *cur = dummy;
		ListNode *h = head;
		while(h){
			if (h->next == nullptr || h->next->val != h->val){
				cur->next = h;
				cur = cur->next;
				h = h->next;
			}else{
				// 越过和h相等的节点
				ListNode *p = h->next;
				while(p!=nullptr && p->val == h->val)p=p->next;
				// 即便找到了也不一定要加到cur后面，因为p可能和之后相等
				h = p;  // p有可能是nullptr
			}
		}
		cur->next = nullptr;
		return dummy->next;
    }
};
int main(){



    return 0;
}
